# CMPS 144

## Assignment 1

**DUE: Friday January 25, 2018 11:59pm**

Turn in the assignment via Canvas.

To write legible answers you will need to be familiar with both [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) and [Latex](https://www.latex-tutorial.com/tutorials/amsmath/)

Before you turn this problem in, make sure everything runs as expected. First, restart the kernel (in the menubar, select Kernel→→Restart) and then run all cells (in the menubar, select Cell→→Run All).

Make sure you fill in any place that says "YOUR CODE HERE" or "YOUR ANSWER HERE", as well as your name below:

In [0]:
NAME = "Amit Saxena"
STUDENT_ID = "1477548"

## Problem 1 

In the famous Monte Hall problem, a contestant is asked to select one of three doors. Behind one of the doors is a new car, while behind the other doors are goats. The game show host (e.g., Monte Hall) opens one of the two remaining doors, revealing a goat. The contestant is then asked if they want to switch the door they picked for the other closed door. Should they switch doors or not?

We can use Bayes' Theorem to decide. Suppose the contestant selects door \#3 (we lose no generality, since we could always renumber the doors so that the contestant did indeed select door \#3). Initially the probabilities of door $i$  hiding the car is   $~~\text{Pr}~(\text{car behind door}~i) = 1/3,~~$  for $i = 1, 2, 3$.

Since we're assuming the contestant picked door \#3, we have the following conditional probabilities, since the host will only open one of the remaining doors (door \#1 or door \#2):

$$\text{Pr}~(\text{host opens door 1  |  door 1 hides the car}) = 0$$
$$\text{Pr}~(\text{host opens door 1  |  door 2 hides the car}) = 1$$
$$\text{Pr}~(\text{host opens door 1  |  door 3 hides the car}) = 1/2$$

$$\text{Pr}~(\text{host opens door 2  |  door 1 hides the car}) = 1$$
$$\text{Pr}~(\text{host opens door 2  |  door 2 hides the car}) = 0$$
$$\text{Pr}~(\text{host opens door 2  |  door 3 hides the car}) = 1/2$$

Let $~C_i~$ denote that door $i$ has the car, and let $~R_j~$ denote that the host has revealed a goat behind door $j$. Then the above conditional probabilities can more succintly be expressed as:

$$\text{Pr}~(R_1~ \vert~ C_1) = 0 \quad\quad \text{Pr}~(R_1~ \vert~ C_2) = 1 \quad\quad \text{Pr}~(R_1~ \vert~ C_3) = 1/2$$

$$\text{Pr}~(R_2~ \vert~ C_1) = 1 \quad\quad \text{Pr}~(R_2~ \vert~ C_2) = 0 \quad\quad \text{Pr}~(R_2~ \vert~ C_3) = 1/2$$

(a)  Use Bayes' theorem to compute: 
$$
\text{Pr}~(\text{door 1 hides the car | host opens door 2})
=\text{Pr}~(C_1~ \vert~ R_2)$$

This equals the probability that the contestant will win the car by switching doors from door \#3 to door \#1, when the host reveals a goat behind door \#2.

HINT -   You need to compute:

$$\text{Pr}~(C_1 ~\vert~ R_2)=\frac{\text{Pr}~(R_2 ~\vert~ C_1)\cdot\text{Pr}~(C_1)}
{\text{Pr}~(R_2 ~\vert~ C_1)\cdot\text{Pr}~(C_1)+
\text{Pr}~(R_2 ~\vert~ C_2)\cdot\text{Pr}~(C_2)+
\text{Pr}~(R_2 ~\vert~ C_3)\cdot\text{Pr}~(C_3)}$$

Simplifying the given expression for $P(C1 | R2)$ we get

$$\frac{2}{3}$$

(b)  Use Bayes' theorem to compute:
$$\text{Pr}~(\text{door 2 hides the car | host opens door 1})
=\text{Pr}~(C_2~ \vert~ R_1)$$

This equals the probability that the contestant will win the car by switching doors from door \#3 to door \#2, when the host reveals a goat behind door \#1.

We can say that $P(C2 | R1)$ = P ( car is on the middle door | host opens the left door)

Now imagine we physically moved door 2 into the left spot (door 1's spot) and we physcally moved door 1 into the middle spot (door 2's old spot).  

Now we can say that 
P ( car is on the middle door | host opens the left door) = $P(C1 | R2)$ 

And we know this is $\frac{2}{3}$

(c)  What is your conclusion? What strategy should the contestant follow, should she switch doors or not (or does it matter?).  What is the probability that the contestant wins a car by following the strategy of ALWAYS switching doors?

The conclusion is that the contestant should always switch doors because the probability they will get the car if they switch is $\frac{2}{3}$ while the probabily they will get the car if they don't switch is $\frac{1}{3}$

## Dataset Overview

For the remaining problems in this assignment we'll use the 500-Person_Gender_Height_Weight_Index dataset. After some preliminary exploration, we'll create a multilinear model to predict the Index from the other factors, and then we'll discover the same model, but using gradient descent. This Index measure consists of a numerical scale:

        0 - extremely weak
        1 - weak
        2 - normal
        3 - overweight
        4 - obese
        5 - extremely obese
    
In this dataset, weight is measured in kilograms, and height is measured in centimeters. At first Index seemed related to the body mass index \[BMI\], so we'll also compute the BMI for comparison.

First, lets read in the dataset (this may take a moment).

In [3]:
# Some useful packages
import numpy as np         # linear algebra
import sklearn as sk       # machine learning
import pandas as pd        # reading in data files, data cleaning
import matplotlib.pyplot as plt   # for plotting
import seaborn as sns      # visualization tool

# This loads in the dataset we'll be using for this assignment.
file_id = '172V7yK-DjsxoaejDlOEl9S1R4A_GdP7U'
link = 'https://drive.google.com/uc?export=download&id={FILE_ID}'
csv_url = link.format(FILE_ID = file_id)

# Import the file as a Pandas data frame.
data = pd.read_csv(csv_url)

# A summary of this dataset:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
Gender    500 non-null object
Height    500 non-null int64
Weight    500 non-null int64
Index     500 non-null int64
dtypes: int64(3), object(1)
memory usage: 15.7+ KB


Notice that this dataset has no Null entries, so our preprocessing will be minimal.  
Let's take a peek at the first records in this dataset.

In [4]:
data.head()  # this command is useful to peek at the dataset after each change

Gender  Height  Weight  Index
0    Male     174      96      4
1    Male     189      87      2
2  Female     185     110      4
3  Female     195     104      3
4    Male     149      61      3

Since we want to include Gender in our initial model, we need to create a new column, Sex, with 0 for female and 1 for male.

In [5]:
# Make a new column, Sex, as described above.  First copy the Gender column to a new Sex column.
data['Sex'] = data['Gender']
  
# Replace strings in Sex, 0 for 'Female' and 1 for 'Male'""
data['Sex'].replace(['Female', 'Male'], [0, 1], inplace=True)

# Look at what we've done.
data.head()

Gender  Height  Weight  Index  Sex
0    Male     174      96      4    1
1    Male     189      87      2    1
2  Female     185     110      4    0
3  Female     195     104      3    0
4    Male     149      61      3    1

## Problem 2 

Add a new column to this dataset, called BMI, for the body mass index. This new column has units of $\frac{\text{[kg]}}{\text{[m]}^2}$. Remember to change units, from cm to m.  

HINT: If you wanted to create a new column named Product, equal to the product of Height and Weight columns, the following would do the trick..

        data['Product'] = data['Height'] * data['Weight']

In [10]:
 data['BMI'] = data['Weight'] / ( (data['Height']/100)**2 )  #  YOUR CODE HERE

# What have you wrought?
data.head()

Gender  Height  Weight  Index  Sex        BMI
0    Male     174      96      4    1  31.708284
1    Male     189      87      2    1  24.355421
2  Female     185     110      4    0  32.140248
3  Female     195     104      3    0  27.350427
4    Male     149      61      3    1  27.476240

Notice that the Index measure isn't the BMI, though it's about 1/10 of the BMI. Clearly Index measures something a bit different than BMI.
    

## Problem 3

Create new columns, 'H' and 'W', to hold the z-transforms of the Height and Weight columns, as we'll be using those transformed values in our model. These new columns measure the number of standard deviations from the mean for each data instance.

Hints:

$\quad\quad \text{data['Height'].mean()}\quad$ computes the mean of the Height column

$\quad\quad \text{data['Height'].std()}\quad$ computes the standard deviation of the Height column.

CORRECTION, not the mean(x)  or the std(x), but the mean and std of the set they belong to.

The z-transform of a random variable x is defined by
$\quad z = \frac{x-\mu}{\sigma}= \frac{x-\text{mean}(x)}{\text{std}(x)}$.

In [17]:
h_mean, h_std = data['Height'].mean() , data['Height'].std()
w_mean, w_std = data['Weight'].mean() , data['Weight'].std()

data['H'] = (data['Height'] - h_mean) / h_std    
data['W'] = (data['Weight'] - w_mean) / w_std    


data.head()

Gender  Height  Weight  Index  Sex        BMI         H         W
0    Male     174      96      4    1  31.708284  0.247691 -0.308808
1    Male     189      87      2    1  24.355421  1.163707 -0.586735
2  Female     185     110      4    0  32.140248  0.919436  0.123523
3  Female     195     104      3    0  27.350427  1.530113 -0.061762
4    Male     149      61      3    1  27.476240 -1.279003 -1.389635

### Training and Test Sets
Next, we'll seperate our data into a training set and a test set. Our model will be based on the training set, and we'll use it to try to predict the values in the test set.  The test set is kept seperate from the training set, so we can get an idea of how well our model will perform on new, as yet unseen data.

In [0]:
# We'll only need the columns 'Sex', 'H', and 'W' for our data, X, and 'Index' for our labels, y.
X = data[['Sex', 'H', 'W']]
y = data['Index']

# This library will help us randomly split our data into training and test sets, 
# using 20% of the data for our test set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=0)

In [19]:
# X holds the data we'll use to predict the labels y.
X_train.head()

Sex         H         W
107    1 -0.668325  1.482277
336    0 -0.179783  1.451396
71     1  0.491962  0.339689
474    0 -0.179783 -0.833781
6      1 -1.401138 -0.432331

In [20]:
# y holds the corresponding labels.
y_train.head()

107    5
336    5
71     4
474    3
6      5
Name: Index, dtype: int64

So we now have X_train, the data we're using to train our model to predict the given corresponding lables y_train.  Once we've trained our model, we can then use it to try to predict the labels for the X_test data (we can then compare our predictions with the known labels stored in y_test).

### Linear Regression

Now we'll create a linear regression model (some would call this multiple regression, since we're using 3 predictor variables rather than 1). The machine learning library SciKit-Learn (sklearn) will do the heavy lifting.

In [21]:
from sklearn import linear_model

# This next line defines a linear model that still must be fit to our training data.
lm = linear_model.LinearRegression()
# Fitting our data to the model is now easy!
model = lm.fit(X_train, y_train)

# Let's examine this model.

# How well does our model predict the sale price?
# The answer to this is called the coefficient of determination, or R^2. This measures the proportion of
# the variation in the data that was explained by our linear model.
print('Training set R^2 =', lm.score(X_train, y_train))
print('Test set R^2 =', lm.score(X_test, y_test))

('Training set R^2 =', 0.8278521846218956)
('Test set R^2 =', 0.816823952075908)


Our linear model did slightly better on the training set than on the test set. This is almost always the case, since our model hadn't seen the test set before.

In [22]:
# The coefficients found by this model are
print('the coefficients =', lm.coef_)
print('the intercept =', lm.intercept_)

('the coefficients =', array([ 0.00208042, -0.5639624 ,  1.09752865]))
('the intercept =', 3.7385774612255336)


These coefficients are used as follows to try to predict the correct Index value:

$$0.00208042\cdot \text{Sex} + (-0.5639624)\cdot H + 1.09752865\cdot W + 3.73857746 = \text{Index}$$

Notice first that the Sex factor only adds only about 0.002 to men's Index scores. This factor contributes so little, it's best left out. Next notice that all else being equal, an increase of one standard deviation in height decreases the resulting index by about 0.56. Also notice that all else being equal, increasing the weight by one standard deviation results in an Index score increase of more than 1.

## Problem 4

Recreate the above linear regression model, but this time not using the Sex column. 

In [34]:
#  YOUR CODE HERE

# We'll only need the columns H', and 'W' for our data, X, and 'Index' for our labels, y.
X = data[['H', 'W']]
y = data['Index']

# This library will help us randomly split our data into training and test sets, 
# using 20% of the data for our test set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=0)



from sklearn import linear_model

# This next line defines a linear model that still must be fit to our training data.
lm = linear_model.LinearRegression()
# Fitting our data to the model is now easy!
model = lm.fit(X_train, y_train)

# Let's examine this model.

# How well does our model predict the sale price?
# The answer to this is called the coefficient of determination, or R^2. This measures the proportion of
# the variation in the data that was explained by our linear model.
print('Training set R^2 =', lm.score(X_train, y_train))
print('Test set R^2 =', lm.score(X_test, y_test))

# The coefficients found by this model are
print('the coefficients =', lm.coef_)
print('the intercept =', lm.intercept_)

old_train_score = 0.8278521846218956
old_test_score = 0.816823952075908

print("new model on training data - old model on training data: {}".format(lm.score(X_train, y_train) - old_train_score))
print("new model on test data - old model on test data: {}".format(lm.score(X_test, y_test) - old_test_score))



('Training set R^2 =', 0.8278516070326984)
('Test set R^2 =', 0.8167131841611407)
('the coefficients =', array([-0.5639695 ,  1.09756578]))
('the intercept =', 3.739566020220715)
new model - train data: -5.77589197137e-07
new model - test data: -0.000110767914767


What is the new accuracy you obtained?  What's the significance of this new model's accuracy, as compared with our previous model?

[YOUR ANSWER HERE]

The new model is only slightly worser than the old model.  The difference in coefficient of determination from the old model to the new model for test data was 0.000110767914767.  

This small degredation in performance makes sense because we did remove a feature dimension from the new model, but the old model showed that this feature had a very small effect, hence the slight degredation in performance.

## Problem 5

Add a column to the dataframe that is the z-transform of the BMI column, and call it zBMI.

In [36]:
bmi_mean, bmi_std = data['BMI'].mean() , data['BMI'].std()

data['zBMI'] = (data['BMI'] - bmi_mean) / bmi_std    

# See what you've done:
data.head()

Gender  Height  Weight  Index  Sex        BMI         H         W      zBMI
0    Male     174      96      4    1  31.708284  0.247691 -0.308808 -0.433708
1    Male     189      87      2    1  24.355421  1.163707 -0.586735 -0.960205
2  Female     185     110      4    0  32.140248  0.919436  0.123523 -0.402777
3  Female     195     104      3    0  27.350427  1.530113 -0.061762 -0.745750
4    Male     149      61      3    1  27.476240 -1.279003 -1.389635 -0.736741

## Problem 6

Redo the regression analysis, with X consisting of the columns H, W, zBMI, and y the Index column. What is this model's accuracy on the training set?

In [38]:
#  YOUR CODE HERE

X = data[['H', 'W', 'zBMI']]
y = data['Index']

# This library will help us randomly split our data into training and test sets, 
# using 20% of the data for our test set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=0)



from sklearn import linear_model

# This next line defines a linear model that still must be fit to our training data.
lm = linear_model.LinearRegression()
# Fitting our data to the model is now easy!
model = lm.fit(X_train, y_train)

# Let's examine this model.

# How well does our model predict the sale price?
# The answer to this is called the coefficient of determination, or R^2. This measures the proportion of
# the variation in the data that was explained by our linear model.
print('Training set R^2 =', lm.score(X_train, y_train))

# The coefficients found by this model are
print('the coefficients =', lm.coef_)
print('the intercept =', lm.intercept_)

old_train_score = 0.8278521846218956

print("new model on training data - old model on training data: {}".format(lm.score(X_train, y_train) - old_train_score))




('Training set R^2 =', 0.8681347833806468)
('the coefficients =', array([-1.36163897,  2.34525421, -1.4924088 ]))
('the intercept =', 3.743046281451959)
new model - train data: 0.0402825987588



The model's coefficient of determination on the training set is 

$$R^2 = 0.8681347833806468$$

This model performed better in the training set than the last 2 models. 




Knowing the z-tranformed BMI has given our model a bit more predictive power. You'll see in Problem 7 that this did somewhat better on the test set too.

## Problem 7

Apply our model to the test set. What is the coefficient of determination of our model on the held-out test set?  This gives us a better idea of how well our model would perform on new, previously unseen data.

In [39]:
# Let's examine this model.

# How well does our model predict the sale price?
# The answer to this is called the coefficient of determination, or R^2. This measures the proportion of
# the variation in the data that was explained by our linear model.
print('Test set R^2 =', lm.score(X_test, y_test))

# The coefficients found by this model are
print('the coefficients =', lm.coef_)
print('the intercept =', lm.intercept_)

old_test_score = 0.816823952075908

print("new model on test data - old model on test data: {}".format(lm.score(X_test, y_test) - old_test_score))


('Test set R^2 =', 0.86223642429136)
('the coefficients =', array([-1.36163897,  2.34525421, -1.4924088 ]))
('the intercept =', 3.743046281451959)
new model on test data - old model on test data: 0.0454124722155


The model's coefficient of determination on the test set is 

$$R^2 = 0.86223642429136$$

This model performed better in the test set than the last 2 models. 

### Gradient Descent

Sometimes with very large datasets, linear regression can't be directly calculated. In such situations and many others, gradient descent is another way to compute the best fit coefficients. For this we'll go back to our model using only the H and W columns to predict the Index value.

Let $X$ be the Height and Weight columns of the dataframe, and let $y$ be the resulting Index. We'll use the loss function 

$$ L(c,b) = \frac{1}{2n}\sum_{i=1}^n \left( y_i-(c^Tx_i + b) \right)^2$$

and use gradient descent to find the minimum of this function, which thus finds the regression coefficients $c$ and the intercept $b$. 

Here $~~~c^T=[h,w]=[\text{height_coefficient}, \text{weight_coefficient]}~~~$ and 
$~~~x_i = 
\begin{bmatrix}
h_i \\
w_i
\end{bmatrix},$

so the above loss function can also be written:

$$ L(h,w,b) = \frac{1}{2n}\sum_{i=1}^n \left( y_i-(h h_i +w w_i+ b) \right)^2$$

$$  \quad\quad= \frac{1}{2n}\sum_{i=1}^n \left( y_i- h h_i - w w_i - b \right)^2$$

The gradient of a function like $L$, written $\nabla L$, has the property that it always points in the direction where $L$ is increasing most rapidly. But we seek the minimum, so we should go in the exact opposite direction, the direction where $L$ is decreasing most rapidly, the direction of the negative gradient, $-\nabla L$.

The $h$ component of $\nabla L$ is the partial derivative of $L$ with respect to $h$, i.e., $\frac{\partial L}{\partial h}$, and similarly for the other two components:

$$\frac{\partial L}{\partial h}=-\frac{1}{n}\sum_{i=1}^n (y_i - hh_i-ww_i-b)\cdot h_i$$

$$\frac{\partial L}{\partial w}=-\frac{1}{n}\sum_{i=1}^n (y_i - hh_i-ww_i-b)\cdot w_i$$

$$\frac{\partial L}{\partial b}=-\frac{1}{n}\sum_{i=1}^n (y_i - hh_i-ww_i-b)\quad~~$$

Thus the gradient of $~L~$ can be expressed as:

$$\nabla L(h,w,b)=
\begin{bmatrix}
\frac{\partial L}{\partial h} \\
\frac{\partial L}{\partial w}  \\
\frac{\partial L}{\partial b} 
\end{bmatrix}=
-\frac{1}{n}
\begin{bmatrix}
\sum_{i=1}^n (y_i - hh_i-ww_i-b)\cdot h_i  \\
\sum_{i=1}^n (y_i - hh_i-ww_i-b)\cdot w_i  \\
\sum_{i=1}^n (y_i - hh_i-ww_i-b)
\end{bmatrix}$$

## Problem 8

Write a Python function that takes as input $~~y, H, W, h, w, b~~$ and outputs the gradient of $~L~$ at the point $~(h,w,b),~$ given the data $~H, W~$.

In [0]:
def delL(y, H, W, h, w, b):
    part_h, part_w, part_b = 0,0,0
    N = len(y)
    
    assert(N == len(H) and N == len(W))
    
    for i in range(N):
      inner = (y[i] - h*H[i] - w*W[i] - b)
      
      part_h += inner * H[i]
      part_w += inner * W[i]
      part_b += inner
    
    part_h *= -(1.0/N)
    part_w *= -(1.0/N)
    part_b *= -(1.0/N)
        
    return (part_h,part_w,part_b)

## Problem 9

Implement gradient descent using your newly forged gradient function. 

In [59]:
y = np.array(data['Index'])
H = np.array(data['H'])
W = np.array(data['W'])

alpha = .05   # the learning rate
num_iterates = 100

# We'll start with the following initial "guess".
h, w, b = .5, .5, 3
print('(h,w,b) =', (h,w,b))

for i in range(num_iterates):
    gradients = delL(y,H,W,h,w,b)
  
    h -= alpha * gradients[0]
    w -= alpha * gradients[1]
    b -= alpha * gradients[2]
    # print('(h,w,b) =', (h,w,b))
    
print('(h,w,b) final =', (h,w,b))

('(h,w,b) =', (0.5, 0.5, 3))
('(h,w,b) final =', (-0.5661947123510837, 1.0869409938138603, 3.743571444143191))
